---
title: "Annotated Seurat Tutorial"
author: "Ziyu Zeng"
date: "2022-09-15"
output: html_document
---

```{r setup, include=FALSE}
# Uncomment if package is not installed
knitr::opts_chunk$set(echo = TRUE)
#install.packages('Seurat')
library(Seurat)
# install.packages('dplyr')
library(dplyr)
# install.packages('patchwork')
library(patchwork)
```

# Clustering Tutorial

Data structure: data is stored in data/ folder.
This tutorial uses single cell data from **10x Genomics** sequenced on the Illumina NextSeq500. 

Raw Data is publicly available [here](https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz). The following codes are based on the [tutorial vignette](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html) on Seurat.

The Read10X function need 3 files from the CellRanger output:
1. matrix.mtx
2. genes.tsv (or features.tsv)
3. barcodes.tsv

We will need the above three files in the same directory for the function to read in data correctly.

***

## Loading data

```{r load dataset, warning=FALSE}
# Load the PBMC dataset. data.dir depends on where you stored the data
pbmc.data <- Read10X(data.dir = "./data/pmbc3k/filtered_gene_bc_matrices/hg19")
# Initialize the Seurat object with the raw (non-normalized data).
# min.cells is a parameter for selecting features(genes) that are expressed in at least 3 cells
# min.features is a parameter for selecting cells that express at least 200 features(genes)
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc

# examine data
dim(pbmc.data)
```

The count matrix has 32738 rows (genes), 2700 columns (cells). Most entries in the matrix is 0 (i.e., a sparse matrix). The values of each entry represent the number of reads of the specific gene for the specific cell.

***

## Quality Control

Common QC metrics:
1. Number of unique genes in each cell
    + Too few -- low quality cells or empty droplets
    + Too many -- cell doublets or multiplets
2. Total number of molecules detected in each cell (similar to 1)
3. % of mitochondrial genes
    + Too high -- low quality cells or dying cells

We use `PercentageFeatureSet()` function to calculate percentage from a set of features, here mitochondrial genes. This function can capture a pattern in gene names in the form of regex (regular expression). For example, 
"^MT-" means all genes starting with MT-. 

For a cheat sheet of regex, see [here](https://www.rexegg.com/regex-quickstart.html).

Filter standards:
  + Unique feature counts is between 200-2500
  + Mitochondrial count <5%

```{r QC, fig.height=7, fig.width=13}
# Calculate mitochondrial gene percentage
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
# Access QC metrics
head(pbmc@meta.data, 5)
# Visualize QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# Visualize as scatter plot
plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt") 
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") 
plot1 + plot2

# Apply filter
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
```

***

## Normalization

The Seurat package provides `NormalizeData()` function for normalization. Default method is "LogNormalize" and default scale factor is 10000.
 
<details>
  <summary>**How does LogNormalize work?**</summary>
  $\text{log}(\frac{\text{feature expression}}{\text{total expression}}\times\text{scale factor})$
</details>
\

```{r normalization}
pbmc <- NormalizeData(pbmc)
```

## Feature Selection

Selecting highly variable features (high cell-to-cell variability) helps to highlight biological signal in single cell datasets. Implemented in the `FindVariableFeatures()` function.

Default setting is returning 2000 features, with "vst" method.

Visualization is achieved by the `VariableFeaturePlot()` function. Can add labels by using the `LabelPoints()` function.

```{r feature, fig.height=5, fig.width=11}
pbmc <- FindVariableFeatures(pbmc)

# Top 10 highly variable features
top10 <- head(VariableFeatures(pbmc), 10)
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1+plot2
```

***

## Data Scaling, Dimension Reduction

Next, we apply a linear transformation ('scaling') that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The `ScaleData()` function:
* Shifts the expression of each gene, so that the mean expression across cells is 0
* Scales the expression of each gene, so that the variance across cells is 1
    + This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
* The results of this are stored in `pbmc[["RNA"]]@scale.data`

By default, ScaleData() uses only the selected features in the previous step. We can explicitly specify the features to be scaled (e.g., all genes), but that will significantly lower the processing speed. The former approach will not affect the clustering result, but may change the heatmap.

```{r scale, eval=FALSE, fig.height=7, fig.width=11}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```

**NEW in Seurat V3:** One-step normalization, scaling and feature selection: `SCTransform`
```{r scale new}
# Make sure you have glmGamPoi installed!!
#if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
# Use glmGamPoi to increase learning speed
#BiocManager::install("glmGamPoi")
pbmc <- SCTransform(pbmc, method = "glmGamPoi", vars.to.regress = "percent.mt", verbose = FALSE)
```

### Dimension reduction through PCA.

```{r pca}
# Run PCA
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
# Examine PCA results
# print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)
# Visualization
# See which genes are included in the principal components 1 and 2
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")
# correlation map, PC1 and PC2
DimPlot(pbmc, reduction = "pca")
# heatmap for pc1
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
# heatmap for pc1-15
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)
```

### Determine dimension cutoff
Slow method: Jack Straw Plot
```{r jackstraw, fig.height=6, fig.width=10, eval=FALSE}
pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)
JackStrawPlot(pbmc, dims = 1:15)
```
Faster: Elbow Plot
```{r elbow}
ElbowPlot(pbmc)
```

Observed elbow is between 9-12. Choose cutoff=10.
For rare cell types, select a higher cutoff. But generally won't affect the clustering results too much.

***

## Clustering, non-linear dimension reduction (UMAP)

Based on K-nearest neighbor (KNN) graph. Can define desired number of dimension using the `dims` argument in `FindNeighbors()`. 

Notes on Resolution:
> The FindClusters() function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the Idents() function.

```{r cluster}
# specify number of PCs in dims= argument
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)
```

UMAP is the more popular method.

```{r umap}
# install umap
#reticulate::py_install(packages ='umap-learn')
pbmc <- RunUMAP(pbmc, dims = 1:10)
DimPlot(pbmc, reduction = "umap", label=TRUE)
# alternatively, use LabelClusters()
# Saving R data for easy reloading/sharing
# saveRDS(pbmc, file = "./output/pbmc_tutorial.rds")
```
Note: Our clustering results is a little different from the original tutorial because we have chosen to use 2000 features instead of all genes. Therefore, labeling will also be different. We will skip the rest of the pipeline for now.
